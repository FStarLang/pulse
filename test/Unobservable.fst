module Unobservable

#lang-pulse
open Pulse

assume val foo : slprop

unobservable
fn incr (x : int)
  preserves foo
  returns int
{
  (x+1)
}

atomic fn test ()
  preserves foo
  returns int
{
  let x = incr 1;
  let x = incr x;
  x
}

unobservable
fn test2 ()
  preserves foo
  returns int
{
  let x = incr 1;
  x
}

[@@expect_failure]
unobservable
fn test3 ()
  preserves foo
  returns int
{
  let x = test ();
  x
}
