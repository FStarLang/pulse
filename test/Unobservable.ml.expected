open Prims
let rec (incr : Prims.int -> Prims.int) = fun x -> x + Prims.int_one
let rec (test : unit -> Prims.int) =
  fun uu___ -> let x = incr Prims.int_one in let x1 = incr x in x1
let rec (test2 : unit -> Prims.int) =
  fun uu___ -> let x = incr Prims.int_one in x
